src/libraries/System.Private.CoreLib/src/System/Enum.cs

```diff
@@ -357,7 +358,21 @@ internal static ulong[] InternalGetValues(RuntimeType enumType)
             ulong[] ulValues = Enum.InternalGetValues(enumType);
             ulong ulValue = Enum.ToUInt64(value);
 
-            return Array.BinarySearch(ulValues, ulValue) >= 0;
+            return FindDefinedIndex(ulValues, ulValue) >= 0;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private static int FindDefinedIndex(ulong[] ulValues, ulong ulValue)
+        {
+            // Binary searching has a higher constant overhead than linear.
+            // For smaller enums, use IndexOf.  For larger enums, use BinarySearch.
+            // This threshold can be tweaked over time as optimizations evolve.
+            const int NumberOfValuesThreshold = 32;
+
+            ref ulong start = ref MemoryMarshal.GetArrayDataReference(ulValues);
```

Comment:
My gut tells me you'll get a slight perf bump if you read `ulValues.Length` and cache it into a local _before_ calling `GetArrayDataReference`, since the JIT can then elide the "is null?" check that `GetArrayDataReference` would normally incur. (The _Length_ getter would have thrown were the value null.)